### ISSUE TITLE
feat: Scalable global reconstruction—Hierarchical, streaming, and memory-efficient techniques (design/research)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/xrpatcher#23]
### Motivation
Global-scale reconstruction from patches (e.g. for climate, weather, satellite imagery) risks exhausting RAM due to accumulator size, patch count, and intermediate operations. This issue synthesizes research, best practices, and practical Python strategies for minimizing memory during reconstruction.

### Current Bottlenecks

The current [`reconstruct` method](https://github.com/jejjohnson/xrpatcher/blob/8b0a6b1056b102ae1849b93af1337bc87050a1bf/xrpatcher/_src/base.py#L136-L262) allocates **three full-size arrays** in RAM:
```python
# BOTTLENECK 1 & 2: Two full-grid arrays (rec + count)
rec_da = xr.DataArray(np.zeros([*new_shape.values()]), dims=dims_labels, coords=coords)
count_da = xr.zeros_like(rec_da)

for ida in tqdm(das):
    da_co = {c: ida[c] for c in coords_labels}
    rec_da.loc[da_co] = rec_da.sel(da_co) + ida * w
    count_da.loc[da_co] = count_da.sel(da_co) + w

# BOTTLENECK 3: Division creates a third full-size array
return rec_da / count_da
```
Additionally, [`get_coords()`](https://github.com/jejjohnson/xrpatcher/blob/8b0a6b1056b102ae1849b93af1337bc87050a1bf/xrpatcher/_src/base.py#L127-L134) eagerly builds N `xr.Dataset` objects (one per patch).

---

#### 1. Streaming Reconstruction to Disk (Zarr/NetCDF)
- Accumulate results directly in a Zarr or NetCDF file chunk-by-chunk on disk.
- Replaces large in-memory arrays with disk-backed accumulators.
- See [Zarr documentation](https://zarr.readthedocs.io/en/stable/python.html) and [scikit-learn patch reconstruction](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.image.reconstruct_from_patches_2d.html).
```python
import zarr

store = zarr.open("reconstruction.zarr", mode="w",
    shape=full_shape, chunks=(1, 360, 720), dtype="float32")
count_store = zarr.open("count.zarr", mode="w",
    shape=full_shape, chunks=(1, 360, 720), dtype="float32")

for idx, patch in enumerate(patch_iterator):
    slices = patcher._get_slices_for_idx(idx)
    store[slices] += patch * weight
    count_store[slices] += weight

# Final division — can also be done chunk-by-chunk
result = zarr.open("reconstruction.zarr", mode="r")[:] / zarr.open("count.zarr", mode="r")[:]
```

#### 2. Patch Iterator Support
- Accept an iterator/generator of patches—not a RAM-hungry list—so only one patch at a time is processed.
- Enables streaming workflows with minimal memory overhead.
```python
# Current (all patches in RAM):
items = [model(patcher[i]) for i in range(len(patcher))]
rec = patcher.reconstruct(items)

# Proposed (streaming, one patch at a time):
def patch_generator():
    for i in range(len(patcher)):
        yield model(patcher[i])

rec = patcher.reconstruct(patch_generator(), output="result.zarr")
```

#### 3. Hierarchical/Multiscale (Pyramid) Reconstruction
- Divide the global domain into tiles (e.g. by lat/lon, time) and reconstruct each subset separately, saving results as you go.
- Reduce peak RAM from full global grid to one tile/grid window.
- Overlaps between tiles should be handled (padding or fractional weights) for border-aware accuracy.
- Composes naturally with xrpatcher's existing [`domain_limits`](https://github.com/jejjohnson/xrpatcher/blob/8b0a6b1056b102ae1849b93af1337bc87050a1bf/xrpatcher/_src/base.py#L59-L62) parameter.
- See [Masked Image Modeling with Local Multi-Scale Reconstruction (CVPR2023)](https://openaccess.thecvf.com/content/CVPR2023/papers/Wang_Masked_Image_Modeling_With_Local_Multi-Scale_Reconstruction_CVPR_2023_paper.pdf).
```
Global Grid (720 × 1440)
├── Tile [0:180, 0:360]    → reconstruct → save to tile_0_0.zarr
├── Tile [0:180, 360:720]  → reconstruct → save to tile_0_1.zarr
├── Tile [180:360, 0:360]  → reconstruct → save to tile_1_0.zarr
├── ...
└── Tile [540:720, 1080:1440] → reconstruct → save to tile_3_3.zarr
    └── Final: xr.open_mfdataset("tile_*.zarr")
```
```python
tiles = [
    dict(lat=slice(0, 180), lon=slice(0, 360)),
    dict(lat=slice(0, 180), lon=slice(360, 720)),
    # ...
]
for tile_limits in tiles:
    tile_patcher = XRDAPatcher(da=data, patches=patches, strides=strides, domain_limits=tile_limits)
    tile_patches = [model(tile_patcher[i]) for i in range(len(tile_patcher))]
    rec_tile = tile_patcher.reconstruct(tile_patches)
    rec_tile.to_zarr(f"tile_{tile_id}.zarr")
    del rec_tile, tile_patches  # free RAM
```

#### 4. In-Place Division
- Use in-place division for the accumulator buffer instead of building extra arrays.
```python
# Current: creates a third full-size array
return rec_da / count_da

# Proposed: in-place, avoids the copy
rec_da.values /= count_da.values
return rec_da
```

#### 5. numpy.memmap
- Leverage numpy's memory mapping to store accumulator arrays on disk efficiently, letting the OS manage paging.
- Zero new dependencies (numpy only).
```python
import numpy as np
import tempfile

rec_data = np.memmap(tempfile.mktemp(), dtype="float32", mode="w+", shape=full_shape)
count_data = np.memmap(tempfile.mktemp(), dtype="float32", mode="w+", shape=full_shape)

for patch, slices in stream:
    rec_data[slices] += patch * weight
    count_data[slices] += weight

rec_data /= count_data
```

#### 6. Lazy Coordinate Generation
- Refactor [`get_coords()`](https://github.com/jejjohnson/xrpatcher/blob/8b0a6b1056b102ae1849b93af1337bc87050a1bf/xrpatcher/_src/base.py#L127-L134) to yield on demand, not build all coordinates at once.
```python
# Current: eagerly builds N xr.Dataset objects
def get_coords(self) -> list[xr.Dataset]:
    coords = []
    for i in range(len(self)):
        coords.append(self[i].coords.to_dataset()[list(self.patches)])
    return coords

# Proposed: lazy generator
def iter_coords(self):
    for i in range(len(self)):
        yield self[i].coords.to_dataset()[list(self.patches)]
```

#### 7. Progressive or Staged Reconstruction
- Progressive/staged reconstruction: work at coarse scale first (downsample), then refine at higher resolutions in later passes.
- [Multi-Stage Progressive Image Restoration](https://arxiv.org/pdf/2102.02808v2)

#### 8. Attention-Based and Grouped Representations
- Use spatial/importance grouping for large sets of patches—update only most relevant subsets.
- See [Memory_Efficient_3D](https://github.com/raj1401/Memory_Efficient_3D) and attention-based models in deep learning.

---

### Related Issues
- Expands and deepens **[#6: Add Scalable Batch Inference Scheme](https://github.com/jejjohnson/xrpatcher/issues/6)** — the hierarchical/pyramid idea originated there.
- Related to **[#21: In-memory caching](https://github.com/jejjohnson/xrpatcher/issues/21)** — caching and reconstruction are complementary performance concerns.
- Related to **[#22: Deep-dive PyTorch DataLoader tutorial](https://github.com/jejjohnson/xrpatcher/issues/22)** — reconstruction at scale is a key section of that tutorial.

### Summary Table

| Technique                        | When to Use            | Python Tools / Tips                                      |
|-----------------------------------|------------------------|----------------------------------------------------------|
| Disk-backed streaming             | Large grids            | Zarr, NetCDF, numpy.memmap                               |
| Patch iterator                    | Always                 | Generator/yield patterns                                 |
| Hierarchical/multiscale           | Multi-region           | domain_limits, recursive tiling                          |
| Progressive/staged                | Deep learning          | Multi-pass/model architectures                           |
| Grouped/attention/importance      | 3D/extreme scale       | Custom slicing, deep models                              |
| In-place division                 | All cases              | numpy, xarray                                           |
| Lazy coords                       | Many patches           | Generators                                              |

### Research References
- [reconstruct_from_patches_2d — scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.image.reconstruct_from_patches_2d.html)
- [python - Large array memory tips](https://stackoverflow.com/questions/14351255/techniques-for-working-with-large-numpy-arrays)
- [Masked Image Modeling with Local Multi-Scale Reconstruction, CVPR2023](https://openaccess.thecvf.com/content/CVPR2023/papers/Wang_Masked_Image_Modeling_With_Local_Multi-Scale_Reconstruction_CVPR_2023_paper.pdf)
- [Multi-Stage Progressive Image Restoration (arXiv)](https://arxiv.org/pdf/2102.02808v2)
- [Memory_Efficient_3D (GitHub)](https://github.com/raj1401/Memory_Efficient_3D)
- [Implementing Efficient Parallel Data Structures on GPUs](https://developer.nvidia.com/gpugems/gpugems2/part-iv-general-purpose-computation-gpus-primer/chapter-33-implementing-efficient)

### Tasks
- [ ] Prototype streaming accumulator (Zarr, NetCDF, memmap)
- [ ] Accept patch generators (not just lists)
- [ ] Implement hierarchical tiling (domain_limits + padding logic)
- [ ] In-place division quick fix
- [ ] Lazy `get_coords()` / `iter_coords()` refactor
- [ ] Profile RAM usage for large reconstructions
- [ ] Document tradeoffs, scenarios, and recommended patterns for users
- [ ] Cross-reference #6, #21, #22
